### PR TITLE
Overlay component

### PR DESCRIPTION
**Pull request by btzr-io**
### Changes
Keep video playing when navigating from a playing video #579
Play / stream videos while searching https://github.com/lbryio/lbry-app/issues/393

### Todo
- [X] Basic ui / animations.
- [X] Fix animations.
- [X] Basic logic.
- [X] Fix scrollbar issue.
- [x] Fix history duplication.
- [x] Cleanup logic.
- [x] Close button that clears the playing URI and hide / close the component.
- [ ] Remove / modify `cancelPlay` in unmount of Video, to keep the playing URI.
- [ ] Improve / fix structure layout and css animations.
- [ ] Swipe to remove ( youtube style )
- [ ] Animate components on `unmount` and `mount`.
- [ ] tidy up / optimize.

~Minor tweaks for video component ( no-controls )~
~Remove / modify `cancelPlay` in unmount of Video, to keep the playing URI~

Check this: https://github.com/lbryio/lbry-app/issues/579#issuecomment-330282722

### CODE REVIEW

File: ui/js/actions/app.js [hunk 1]
```diff
@@ -219,3 +219,15 @@ export function doChangeVolume(volume) {
     });
   };
 }
+
+export function doOpenOverlayMedia() {
```
Comment (kauffj) [inline ui/js/actions/app.js]: I suspect this code can be written without this action. We can know whether or not it should be open solely from existing app state.

File: ui/js/actions/app.js [hunk 2]
```diff
@@ -219,3 +219,15 @@ export function doChangeVolume(volume) {
...
+  return {
+    type: types.OPEN_OVERLAY_MEDIA,
+  };
+}
+
+export function doCloseOverlayMedia() {
```
Comment (kauffj) [inline ui/js/actions/app.js]: I also suspect this action can be avoided, and instead just be `doSetPlayingUri(null)`.

File: ui/js/component/app/view.jsx [hunk 3]
```diff
@@ -53,6 +54,7 @@ class App extends React.PureComponent {
         <Theme />
         <Header />
         <div id="main-content">
+          <OverlayRouter />
```
Comment (kauffj) [inline ui/js/component/app/view.jsx]: This should not be a router. This should just be `VideoOverlay`.

File: ui/js/component/overlayRouter/view.jsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+  render() {
+    const { currentPage, params, playingUri, keepPlaying } = this.props;
+
+    const uri = params.uri || playingUri;
+    const isPage = currentPage === "show";
+
+    const page = uri && (keepPlaying || isPage)
+      ? <ShowPage expanded={isPage} uri={uri} />
```
Comment (kauffj) [inline ui/js/component/overlayRouter/view.jsx]: The `ShowPage` should not be doing double duty of being both a full page and the overlay component. Allowing them to each be their own function will be a much cleaner design.

File: ui/scss/component/_overlay-media.scss [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  &.minimized {
+    width: 100% !important;
+    position: fixed;
+    top: 0;
+    left: 0;
+    /*bottom: $inner-height + $inner-margin;*/
+    transform: translateY(calc(100vh - 248px));
+    z-index: 99;
```
Comment (kauffj) [inline ui/scss/component/_overlay-media.scss]: Right now our z-index is very clean:

- 3 layers (1, 2, 3) in the app.
- 2 layers (9999, 10000) for modals and snackbars, respectively.

I suspect z-index: 3 here would be fine, or at most 4.

Comment (btzr-io) [reply ui/scss/component/_overlay-media.scss]: Ok, I need rewrite styles anyway :+1:

**Comment by kauffj** [comment]
Noooooooooooooooooo it was so close!

**Pull request closed without merging**